scripts/datastore_upgrade.py

```diff
@@ -0,0 +1,422 @@
+  Returns:
+    True on success, False otherwise.
+  """
+  logging.debug("Process entity {}".format(str(entity)))
+  key = entity.keys()[0]
+  one_entity = entity[key][APP_ENTITY_SCHEMA[0]]
+  version = entity[key][APP_ENTITY_SCHEMA[1]]
+
+  app_id = entity_utils.get_prefix_from_entity_key(key)
+  validated_entity = ds_distributed.validated_result(app_id, entity)
+
+  is_tombstone = validated_entity[key][APP_ENTITY_SCHEMA[0]] == datastore_server.TOMBSTONE
+  if not validated_entity or is_tombstone:
+    return delete_entity_from_table(key, datastore)
+
+  if not (one_entity == validated_entity[key][APP_ENTITY_SCHEMA[0]]
+      and version == validated_entity[key][APP_ENTITY_SCHEMA[1]]):
+    return update_entity_in_table(key, validated_entity, datastore)
+
+  return True
+
+def update_entity_in_table(key, validated_entity, datastore):
+  """ Updates the APP_ENTITY_TABLE with the valid entity.
+  Args:
+    key: A str representing the row key to update in the table.
+    validated_entity: A validated entity which needs to be updated in place of
+    the current entity.
+    datastore: A reference to the batch datastore interface.
+  """
+  datastore.batch_put_entity(APP_ENTITY_TABLE, [key], APP_ENTITY_SCHEMA, validated_entity[key])
+
+def delete_entity_from_table(key, datastore):
+  """ Performs a hard delete on the APP_ENTITY_TABLE for the given row key.
+  Args:
+    key: A str representing the row key to delete from the table.
+    datastore: A reference to the batch datastore interface.
+  """
+  datastore.batch_delete(APP_ENTITY_TABLE, [key])
+
+def stop_cassandra(db_ips, status_dict, keyname):
+  """ Stops Cassandra.
+  Args:
+    db_ips: A list of database node IPs to stop Cassandra on.
+    status_dict: A dictionary to record the status of the executed process.
+    keyname: A string containing the deployment's keyname.
+  """
+  logging.info("Stopping Cassandra...")
+  for ip in db_ips:
+    stop_service_cmd = STOP_SERVICE_SCRIPT + CASSANDRA_WATCH_NAME
+    cmd_status = utils.ssh(ip, keyname, stop_service_cmd)
+    stop_cassandra_ip = STOP_CASSANDRA + ":" + ip
+    if not cmd_status == 0:
+      logging.error("Monit was unable to stop Cassandra.")
+      status_dict[stop_cassandra_ip] = FAILURE
+      sys.exit(1)
+    logging.info("Successfully stopped Cassandra.")
+    status_dict[stop_cassandra_ip] = SUCCESS
+
+def stop_zookeeper(zk_ips, status_dict, keyname):
+  """ Stops ZooKeeper.
+  Args:
+    zk_ips: A list of zookeeper node IPs to stop ZooKeeper on.
+    status_dict: A dictionary to record the status of the executed process.
+    keyname: A string containing the deployment's keyname.
+  """
+  logging.info("Stopping ZooKeeper...")
+  for ip in zk_ips:
+    stop_service_cmd = STOP_SERVICE_SCRIPT + ZK_WATCH_NAME
+    cmd_status = utils.ssh(ip, keyname, stop_service_cmd)
+    stop_zookeeper_ip = STOP_ZOO_KEEPER + ":" + ip
+    if not cmd_status == 0:
+      logging.error("Monit was unable to stop ZooKeeper.")
```

Comment:
I changed this to stop all ZooKeeper nodes regardless of error in stopping any. It is not showing up on the diff because its two lines below.